pkg/parse/events/timed_event_producer.go

```diff
@@ -0,0 +1,126 @@
+	// the namespace-controller wants to trigger a resync.
+	// TODO: Use a channel, instead of a timer checking a locked variable.
+	NamespaceControllerPeriod time.Duration
+	// RetryBackoff is how long the Parser waits between retries, after an error.
+	RetryBackoff wait.Backoff
+}
+
+// Generate generates and sends events to the HandleEventFunc based on the
+// configured time periods, until the context is done.
+// Generate implements GenerateFunc.
+func (t *TimedEventProducer) Generate(ctx context.Context, handleFn HandleEventFunc) {
+	// Use timers, not tickers.
+	// Tickers can cause memory leaks and continuous execution, when execution
+	// takes longer than the tick duration.
+	runTimer := time.NewTimer(t.PollingPeriod)
+	defer runTimer.Stop()
+
+	resyncTimer := time.NewTimer(t.ResyncPeriod)
+	defer resyncTimer.Stop()
+
+	statusUpdateTimer := time.NewTimer(t.StatusUpdatePeriod)
+	defer statusUpdateTimer.Stop()
+
+	nsEventTimer := time.NewTimer(t.NamespaceControllerPeriod)
+	defer nsEventTimer.Stop()
+	// Disable namespace events when period is zero/default.
+	if t.NamespaceControllerPeriod == 0 {
+		nsEventTimer.Stop()
+	}
+
+	backoff := copyBackoff(t.RetryBackoff)
+	retryLimit := backoff.Steps
+	retryTimer := time.NewTimer(backoff.Duration)
+	defer retryTimer.Stop()
+
+	resetRetryBackoff := func() {
+		backoff = copyBackoff(t.RetryBackoff)
+		retryTimer.Reset(backoff.Duration)
+	}
+
+	var result HandleResult
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-resyncTimer.C:
+			result = handleFn(&ResyncEvent{})
+
+			// Schedule next resync attempt
+			resyncTimer.Reset(t.ResyncPeriod)
+
+		case <-runTimer.C:
+			result = handleFn(&SyncEvent{})
+
+			// Schedule next sync attempt
+			runTimer.Reset(t.PollingPeriod)
+
+		case <-retryTimer.C:
+			if backoff.Steps == 0 {
+				klog.Infof("Retry limit (%v) has been reached", retryLimit)
+				// Don't reset retryTimer if retry limit has been reached.
+				continue
+			}
+
+			event := &RetryEvent{
+				ResetRetryBackoff: resetRetryBackoff,
+			}
+
+			retryDuration := backoff.Step()
+			retries := retryLimit - backoff.Steps
+			klog.Infof("a retry is triggered (retries: %v/%v)", retries, retryLimit)
+
+			result = handleFn(event)
+
+			// Schedule next retry attempt
+			retryTimer.Reset(retryDuration)
+
+		case <-statusUpdateTimer.C:
+			result = handleFn(&StatusEvent{})
+
+			// Schedule next status update attempt
+			statusUpdateTimer.Reset(t.StatusUpdatePeriod)
+
+		case <-nsEventTimer.C:
+			result = handleFn(&NamespaceResyncEvent{})
+
+			// Schedule next namespace event check attempt
+			nsEventTimer.Reset(t.NamespaceControllerPeriod)
+		}
+
+		// If the RunFunc was called, it will likely have updated the status.
+		// So we can delay the next StatusEvent.
+		// If not, the event was probably handled very quickly, so we don't want
+		// to delay the StatusEvent, otherwise it may never trigger.
+		if !result.Skipped {
```

Comment:
Might want to pull this out to a different PR, since it's actually a bug fix for the namespace event handling.